Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp

```diff
@@ -318,7 +318,23 @@ SkPaint GraphicsContextSkia::createFillPaint(std::optional<Color> fillColor) con
         if (globalAlpha < 1)
             shadowColor = shadowColor.colorWithAlphaMultipliedBy(globalAlpha);
         auto [r, g, b, a] = shadowColor.toColorTypeLossy<SRGBA<uint8_t>>().resolved();
-        paint.setImageFilter(SkImageFilters::DropShadow(shadow->offset.width(), shadow->offset.height(), sigma, sigma, SkColorSetARGB(a, r, g, b), nullptr));
+
+        sk_sp<SkImageFilter> filter;
+        switch (shadowStyle) {
+        case ShadowStyle::Outset:
+            filter = SkImageFilters::DropShadow(shadow->offset.width(), shadow->offset.height(), sigma, sigma, SkColorSetARGB(a, r, g, b), nullptr);
+            break;
+        case ShadowStyle::Inset: {
+            auto dropShadow = SkImageFilters::DropShadowOnly(shadow->offset.width(), shadow->offset.height(), sigma, sigma, SK_ColorBLACK, nullptr);
+            filter = SkImageFilters::ColorFilter(SkColorFilters::Blend(SkColorSetARGB(a, r, g, b), SkBlendMode::kSrcOut), dropShadow);
+            break;
+        }
+        case ShadowStyle::None:
+            RELEASE_ASSERT_NOT_REACHED();
+            break;
+        }
+
+        paint.setImageFilter(filter);
```

Comment:
I see in the cairo case we have an isVisible() to avoid creating the shadow if the color is fully transparent, for example. I also wonder what happens if shadow ignores transforms, we don't handle that case here either, no?

Reply:
Applied the same optimization to Skia, and added a FIXME for the ignore transform case